### ISSUE TITLE
[css-align-3][css-flexbox-1] How to determine the last baseline of a flex container with different alignment groups.

### ISSUE THREAD

**Issue by bfgeek** [w3c/csswg-drafts#7641]
See: https://www.software.hixie.ch/utilities/js/live-dom-viewer/?saved=10608
```html
<!DOCTYPE html>
<div style="display: inline-flex; align-items: last baseline;">
  <div style="background: lime; font-size: 30px;">line1<br>line2</div>
  <div style="display: flex;">
    <div style="font-size: 30px; align-self: last baseline; background: cyan;">line1<br>line2<br>line3</div> 
    <div style="font-size: 80px; align-self: baseline; background: hotpink;">line1<br>line2</div> 
  </div>
</div>
```
https://drafts.csswg.org/css-flexbox-1/#flex-baselines

Should the last baseline of the "lime" box be aligned with the last-baseline of the "cyan" box, or the "hotpink" box?

The rules just look at if something "participates in baseline alignment" so by that rule it should be the "hotpink" box.
But likely should be the "cyan" box.

**Comment by tabatkins** [comment]
Yeah, I *think* this is just a bit of behavior that slipped thru the edits when we upgraded from "baseline" to "first/last baseline". I suspect we should prefer taking from the last-baseline element when determining the inner flexbox's last baseline, and mutatis mutandis for the first.

This does mean that the last baseline can potentially be higher than the first baseline, tho (say, both items are single-line, and the last-baseline one has a smaller font-size). Is that ok?

**Comment by bfgeek** [comment]
> Is that ok?
I think its ok - but reserve the right to complain later if it isn't. :)

**Comment by tabatkins** [comment]
Okay, so suggested edits are to modify Step 1 of <https://drafts.csswg.org/css-flexbox-1/#flex-baselines> to first look at items with the same baseline alignment preference as the requested baseline, falling back to *any* baseline-aligned item if there aren't any matching.

**Comment by bfgeek** [comment]
I think I have a slight preference for look at items which match the baseline alignment preference - then fallback to any item.

Otherwise we have three(!) levels of fallback - e.g.
1. Baseline alignment preference.
2. Any baseline alignment.
3. Any.
vs.
1. Baseline alignment preference.
2. Any.

The two stage fallback would be web compatible with what we have today.

**Comment by fantasai** [comment]
@bfgeek Supposing we want to last-baseline align a table or grid, and it has first-baseline aligned content in its last row... I can see the argument going either way wrt falling back three levels or two. Consider the case of having what are, essentially, one-line items first-baseline aligned to each other but the first item is a replaced element; wouldn't we rather take the real baseline than the synthesized one from the replaced item?

**Comment by bfgeek** [comment]
> Consider the case of having what are, essentially, one-line items first-baseline aligned to each other but the first item is a replaced element; wouldn't we rather take the real baseline than the synthesized one from the replaced item?

For this case it wouldn't really matter though as currently spec'd we'd take the last-baseline from the last item?

**Comment by bfgeek** [comment]
The general problem that I think you are hinting at already exists without replaced elements, e.g.

```html
<div style="display: flex; align-items: last baseline;">
  <div>baseline</div>
  <div style="display: flex; align-items: baseline;">
    <div>line1<br>line2<br>line3</div>
    <div>baseline</div>
  </div>
</div>
```

(FF rendering is badly wrong here) but here "baseline" "line1" "baseline" would all be aligned.

**Comment by fantasai** [comment]
While it's clear we didn't think through the case of having both first- and last-baseline-aligned items in the flex line, IIRC it was pretty intentional that if only one of them existed we took that shared baseline.

i.e. in 

> If any of the flex items on the flex container’s startmost/endmost flex line participate in baseline alignment, the flex container’s first/last main-axis baseline set is generated from the shared alignment baseline of those flex items.

The fact that we wrote “participate in baseline alignment” and not “participate in first/last baseline alignment” was, insofar as I can remember, intentional.

This implies that, for consistency, the mixed case should be resolved with 3-way fallback (desired baseline alignment, any baseline alignment, then just grab the baselines from an arbitrary non-aligned element).

**Comment by tabatkins** [comment]
(Not to say we can't dispute that and go with something simpler; this was just the intention when we wrote it originally.)

**Comment by bfgeek** [comment]
This is an advantage of how the its specified in tables - in that you end up picking the lowest vs. the "first/last" item (look at all items - pick the lowest baseline).

**Comment by fantasai** [comment]
The resolution got recorded as we finally take the first baseline of the first item, but we also had discussed taking the highest first baseline... is the resolution recorded correctly, or did we actually prefer @bfgeek's suggestion to take the highest one? Ian, do you happen to remember?

**Comment by bfgeek** [comment]
I believe we take the "highest" one - at least I remember we had a conversation about the "overflow:scroll;flex-direction:column-reverse;" case and thought the clamping to the top in that case was likely acceptable.

E.g. view https://www.software.hixie.ch/utilities/js/live-dom-viewer/?saved=10927 on chrome canary.

(This is also what I added as tests relatively soon after this discussion).

Ian

**Comment by fantasai** [comment]
@tabatkins and I have drawn up a PR for this at https://github.com/w3c/csswg-drafts/pull/8341

We'd like to get confirmation from the WG on a few points, and also wanted to ask @bfgeek for review. :)
